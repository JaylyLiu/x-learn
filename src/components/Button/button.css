.btn {
  margin: 0;
  padding: 0;
  min-width: 80px;
  font-weight: 500;
  text-align: center;
  border: none;
  border-radius: 3px;
  transition: background 0.15s;
  border-radius: 3px;
  line-height: 1;
  text-transform: uppercase;
}

button:focus {
  outline: none;
}

button:hover:enabled {
  cursor: pointer;
}

button:disabled,
button[disabled],
button-disabled {
  opacity: 0.5;
  cursor: not-allowed;
}

/* shadowed */
.btn-shadowed {
  box-shadow: 0 2px 8px 0 rgba(0, 0, 0, 0.25);
}

/* Button color */
.btn-main {
  color: #ffffff;
  background: #00BFBD;
}

.btn-main:hover:enabled {
  background: rgba(0, 191, 188, 0.377);
}

.btn-default {
  color: #333333;
  background: #eeeeee;
}

.btn-default:hover:enabled {
  background: #a7a7a7;
}

.btn-error {
  color: #ffffff;
  background: #cb242b;
}

.btn-error:hover:enabled {
  background: rgba(0, 191, 188, 0.377);
}

.btn-success {
  color: #ffffff;
  background: #5ebd5e;
}

.btn-success:hover:enabled {
  background: #428442;
}

.btn-instant-tutoring {
  color: #ffffff;
  background: #5ebfe2;
}

.btn-instant-tutoring:hover:enabled {
  background: #42869e;
}

.btn-gradient {
  color: #ffffff;
  background-image: linear-gradient(to right, #ff4338, #ff786b);
}

.btn-gradient:hover:enabled {
  background-image: linear-gradient(rgba(0, 0, 0, 0.3), rgba(0, 0, 0, 0.3)),
    linear-gradient(to right, #ff685f, #ffac7b);
}

/* ghost */
.btn-ghost-main {
  font-weight: 400;
  color: #00BFBD;
  background: transparent;
  border: solid 1px #00BFBD;
}

.btn-ghost-main:hover:enabled {
  /* color: #b32e27;
  border: solid 1px #b32e27; */
  opacity: 0.6;
}

.btn-ghost-default {
  font-weight: 400;
  color: #333333;
  background: transparent;
  border: solid 1px #333333;
}

.btn-ghost-default:hover:enabled {
  /* color: #a7a7a7;
  border: solid 1px #a7a7a7; */
  opacity: 0.6;
}

.btn-ghost-error {
  font-weight: 400;
  color: #cb242b;
  background: transparent;
  border: solid 1px #cb242b;
}

.btn-ghost-error:hover:enabled {
  /* color: #00BFBD;
  border: solid 1px #00BFBD; */
  opacity: 0.6;
}

.btn-ghost-success {
  color: #5ebd5e;
  background: transparent;
  border: solid 1px #5ebd5e;
}

.btn-ghost-success:hover:enabled {
  /* color: #428442;
  border: solid 1px #428442; */
  opacity: 0.6;
}

.btn-ghost-instant-tutoring {
  color: #5ebfe2;
  background: transparent;
  border: solid 1px #5ebfe2;
}

.btn-ghost-instant-tutoring:hover:enabled {
  /* color: #42869e;
  border: solid 1px #42869e; */
  opacity: 0.6;
}
.btn-ghost-pure {
  color: #ffffff;
  background: transparent;
  border: solid 1px #ffffff;
}

.btn-ghost-pure:hover:enabled {
  opacity: 0.7;
}

/* Button size */
.btn-big {
  min-height: 42px;
  padding: 8px 12px;
  font-size: 14px;
}

.btn-standard {
  min-height: 36px;
  padding: 10px 12px;
  font-size: 14px;
}

.btn-medium {
  min-height: 32px;
  padding: 8px 12px;
  font-size: 14px;
}

.btn-small {
  min-height: 32px;
  padding: 9px 12px;
  font-size: 12px;
}

/* Action Button Size*/
.btn-action-standard {
  border-radius: 4px;
  min-height: 42px;
  padding: 10px 50px;
  font-size: 18px;
  min-width: 200px;
}
.btn-action-medium {
  border-radius: 4px;
  min-height: 60px;
  padding: 10px 50px;
  font-size: 20px;
  min-width: 200px;
}
.btn-action-large {
  border-radius: 4px;
  min-height: 80px;
  padding: 10px 50px;
  font-size: 25px;
  min-width: 200px;
}

.btn-loading {
  position: relative;
}
.btn-loading * {
  visibility: hidden;
}
.btn-loading:before {
  width: 18px;
  height: 18px;
  margin: auto;
  content: '';
  position: absolute;
  animation: rotate 0.8s infinite linear;
  border: 2px solid #fff;
  border-right-color: transparent;
  border-radius: 50%;
  top: 0;
  bottom: 0;
  left: 0;
  right: 0;
  z-index: 100;
}
.btn-loading:after {
  width: 100%;
  height: 100%;
  content: '';
  position: absolute;
  top: 0;
  bottom: 0;
  left: 0;
  right: 0;
  background: #00BFBD;
  z-index: 10;
  border-radius: 3px;
}

.btn-ghost-main.btn-loading:before,
.btn-ghost-default.btn-loading:before,
.btn-ghost-pure.btn-loading:before {
  width: 18px;
  height: 18px;
  margin: auto;
  content: '';
  position: absolute;
  animation: rotate 0.8s infinite linear;
  border: 2px solid #00BFBD;
  border-right-color: transparent;
  border-radius: 50%;
  top: 0;
  bottom: 0;
  left: 0;
  right: 0;
  z-index: 100;
}
.btn-ghost-main.btn-loading:after,
.btn-ghost-default.btn-loading:after,
.btn-ghost-pure.btn-loading:after {
  width: 100%;
  height: 100%;
  content: '';
  position: absolute;
  top: 0;
  bottom: 0;
  left: 0;
  right: 0;
  background: transparent;
  z-index: 10;
  border-radius: 3px;
}
.btn-ghost-default.btn-loading:before {
  border: 2px solid #333333;
  border-right-color: transparent;
}
.btn-ghost-pure.btn-loading:before {
  border: 2px solid #ffffff;
  border-right-color: transparent;
}

@keyframes rotate {
  0% {
    transform: rotate(0deg);
  }
  100% {
    transform: rotate(360deg);
  }
}
