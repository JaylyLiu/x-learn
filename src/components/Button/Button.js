import React, { Component } from 'react';
import PropTypes from 'prop-types';
import cx from 'classnames';
import './button.css';

class Button extends Component {
  static propTypes = {
    classnames: PropTypes.string.isRequired,
    children: PropTypes.node.isRequired,
  };

  stopPropagation = evt => {
    evt = evt || window.event; //获取event对象
    if (evt && evt.stopPropagation)
      //因此它支持W3C的stopPropagation()方法
      evt.stopPropagation();
    else {
      //否则，我们需要使用IE的方式来取消事件冒泡
      window.event.cancelBubble = true;
    }
  };

  render() {
    const {
      classnames,
      children,
      disabled,
      style,
      onClick,
      id,
      href,
      target,
      classNameWrap,
    } = this.props;
    const myButton = (
      <button
        id={id}
        type="button"
        className={classnames}
        disabled={disabled}
        style={style}
        onClick={onClick}>
        {children}
      </button>
    );

    if (href && !disabled) {
      return (
        <a
          href={href}
          target={target || '_self'}
          className={classNameWrap}
          onClick={this.stopPropagation}>
          {myButton}
        </a>
      );
    } else {
      return myButton;
    }
  }
}

const withStyle = WrappedComponent =>
  class extends Component {
    static propTypes = {
      iistyle: PropTypes.oneOf([
        'default',
        'main',
        'success',
        'error',
        'instant-tutoring',
        'pure',
      ]),
      size: PropTypes.oneOf(['big', 'standard', 'medium', 'small']),
    };

    static defaultProps = {
      iistyle: 'default',
      size: 'standard',
    };

    state = {
      loading: false,
    };

    onClick = async event => {
      if (this.props.async) {
        this.setState({ loading: true });
        await this.props.onClick(event);
        this.setState({ loading: false });
      } else {
        this.props.onClick && this.props.onClick(event);
      }
    };

    render() {
      const {
        className,
        ghost,
        loading,
        iistyle,
        shadowed,
        size,
        gradient,
        action,
        ...rest
      } = this.props;
      const classnames = cx(className, 'btn', `btn-${size}`, {
        [`btn-${iistyle}`]: !ghost,
        [`btn-ghost-${iistyle}`]: ghost,
        'btn-shadowed': shadowed,
        'btn-gradient': gradient,
        [`btn-action-${size}`]: action,
        'btn-loading': loading || this.state.loading,
      });
      return (
        <WrappedComponent
          {...rest}
          disabled={this.props.disabled || loading || this.state.loading}
          classnames={classnames}
          onClick={this.onClick}
        />
      );
    }
  };

export default withStyle(Button);
