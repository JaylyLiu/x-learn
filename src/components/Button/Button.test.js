import React from 'react';
import renderer from 'react-test-renderer';
import { shallow } from 'enzyme';
import Button from './Button';

describe('Button', () => {
  it('renders correctly', () => {
    const tree = renderer.create(<Button>Default</Button>).toJSON();
    expect(tree).toMatchSnapshot();
  });

  it('renders main button correctly', () => {
    const tree = renderer.create(<Button iistyle="main">Main</Button>).toJSON();
    expect(tree).toMatchSnapshot();
  });

  it('renders error button correctly', () => {
    const tree = renderer
      .create(<Button iistyle="error">Error</Button>)
      .toJSON();
    expect(tree).toMatchSnapshot();
  });

  it('renders success button correctly', () => {
    const tree = renderer
      .create(<Button iistyle="success">Success</Button>)
      .toJSON();
    expect(tree).toMatchSnapshot();
  });

  it('renders instant-tutoring button correctly', () => {
    const tree = renderer
      .create(<Button iistyle="instant-tutoring">Instant Tutoring</Button>)
      .toJSON();
    expect(tree).toMatchSnapshot();
  });

  it('renders big size button correctly', () => {
    const tree = renderer.create(<Button size="big">Big size</Button>).toJSON();
    expect(tree).toMatchSnapshot();
  });

  it('renders standard size button correctly', () => {
    const tree = renderer
      .create(<Button size="standard">Standard size</Button>)
      .toJSON();
    expect(tree).toMatchSnapshot();
  });

  it('renders medium size button correctly', () => {
    const tree = renderer
      .create(<Button size="medium">Medium size</Button>)
      .toJSON();
    expect(tree).toMatchSnapshot();
  });

  it('renders small size button correctly', () => {
    const tree = renderer
      .create(<Button size="small">Small size</Button>)
      .toJSON();
    expect(tree).toMatchSnapshot();
  });

  it('has click event', () => {
    const fn = jest.fn();
    const wrapper = shallow(<Button onClick={fn}>Click</Button>).dive();
    wrapper.find('button').simulate('click');
    expect(fn).toBeCalled();
  });

  it('disabled', () => {
    const fn = jest.fn();
    const wrapper = shallow(<Button disabled>Click</Button>).dive();
    wrapper.find('button').simulate('click');
    expect(fn).not.toBeCalled();
  });
});
