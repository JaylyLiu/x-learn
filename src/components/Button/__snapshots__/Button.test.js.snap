// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Button renders big size button correctly 1`] = `
<button
  className="btn btn-big btn-default"
  disabled={false}
  onClick={[Function]}
  type="button"
>
  Big size
</button>
`;

exports[`Button renders correctly 1`] = `
<button
  className="btn btn-standard btn-default"
  disabled={false}
  onClick={[Function]}
  type="button"
>
  Default
</button>
`;

exports[`Button renders error button correctly 1`] = `
<button
  className="btn btn-standard btn-error"
  disabled={false}
  onClick={[Function]}
  type="button"
>
  Error
</button>
`;

exports[`Button renders instant-tutoring button correctly 1`] = `
<button
  className="btn btn-standard btn-instant-tutoring"
  disabled={false}
  onClick={[Function]}
  type="button"
>
  Instant Tutoring
</button>
`;

exports[`Button renders main button correctly 1`] = `
<button
  className="btn btn-standard btn-main"
  disabled={false}
  onClick={[Function]}
  type="button"
>
  Main
</button>
`;

exports[`Button renders medium size button correctly 1`] = `
<button
  className="btn btn-medium btn-default"
  disabled={false}
  onClick={[Function]}
  type="button"
>
  Medium size
</button>
`;

exports[`Button renders small size button correctly 1`] = `
<button
  className="btn btn-small btn-default"
  disabled={false}
  onClick={[Function]}
  type="button"
>
  Small size
</button>
`;

exports[`Button renders standard size button correctly 1`] = `
<button
  className="btn btn-standard btn-default"
  disabled={false}
  onClick={[Function]}
  type="button"
>
  Standard size
</button>
`;

exports[`Button renders success button correctly 1`] = `
<button
  className="btn btn-standard btn-success"
  disabled={false}
  onClick={[Function]}
  type="button"
>
  Success
</button>
`;
